lib/stormpath-sdk/auth/register_service_provider.rb

```diff
@@ -0,0 +1,46 @@
+module Stormpath
+  module Authentication
+    class RegisterServiceProvider
+      attr_reader :client, :identity_provider, :assertion_consumer_service_url,
+                  :entity_id, :name, :description, :name_id_format
+
+      def initialize(client, identity_provider, assertion_consumer_service_url, entity_id, options = {})
+        @client = client
+        @identity_provider = identity_provider
+        @assertion_consumer_service_url = assertion_consumer_service_url
+        @entity_id = entity_id
+        @name = options[:name] || nil
+        @description = options[:description] || nil
+        @name_id_format = options[:name_id_format] || nil
+      end
+
+      def call
+        map_identity_provider_and_registered_service_provider
+        registered_service_provider
+      end
+
+      private
+
+      def map_identity_provider_and_registered_service_provider
+        identity_provider.saml_service_provider_registrations.create(
+          service_provider: { href: registered_service_provider.href }
+        )
+      end
+
+      def registered_service_provider
+        @registered_service_provider ||=
+          client.registered_saml_service_providers.create(registered_service_provider_params)
+      end
+
+      def registered_service_provider_params
+        {}.tap do |body|
+          body[:assertion_consumer_service_url] = assertion_consumer_service_url
+          body[:entity_id] = entity_id
+          body[:name] = name
+          body[:description] = description
+          body[:name_id_format] = name_id_format
+        end.compact
+      end
+    end
+  end
+end
```

Comment:
In the current state we are using this as follows

```ruby
identity_provider = application.saml_policy.identity_provider
identity_provider.register_service_provider(
  assertion_consumer_service_url: assertion_consumer_service_url,
  entity_id: entity_id
)
```

if me move the temp class we have

```ruby
application.saml_policy.identity_provider.identity_provider.register_service_provider(
  assertion_consumer_service_url: assertion_consumer_service_url,
  entity_id: entity_id
)

```
I did several refactorings in this class, each one was breaking law of Demeter. So in my opinion register service provider should be called directly on application

```ruby
application.register_service_provider(
  assertion_consumer_service_url: assertion_consumer_service_url,
  entity_id: entity_id
)

```
and in the application we would have

```ruby
def register_service_provider(options = {})
  Stormpath::Authentication::RegisterServiceProvider.new(saml_policy.identity_provider, options).call
end
```

to me this seems the right way to do it. The application class is getting bigger and probably needs some cleanup(probably with modules that is with composition) but this is not the reason to avoid putting methods there which are obviously belonging there.

Cilim and I paired and refactored this so  he will make a new PR